systemtest/sourcemap_test.go

```diff
@@ -0,0 +1,344 @@
+	uploadSourcemap(t, srv, "../testdata/sourcemap/bundle.js.map",
+		"http://localhost:8000/test/e2e/general-usecase/bundle.js.map",
+		"apm-agent-js",
+		"1.0.0",
+	)
+	systemtest.Elasticsearch.ExpectDocs(t, "apm-*-sourcemap", nil)
+
+	// trigger cache
+	systemtest.SendRUMEventsPayload(t, srv, "../testdata/intake-v2/errors_rum.ndjson")
+	systemtest.Elasticsearch.ExpectDocs(t, "apm-*-error", nil)
+
+	// delete sourcemap index
+	deleteIndex(t, "apm-*-sourcemap*")
+
+	// wait for the cache to expire
+	time.Sleep(time.Second)
+
+	// no sourcemap available, since the index was removed and the cache was expired
+	systemtest.SendRUMEventsPayload(t, srv, "../testdata/intake-v2/errors_rum.ndjson")
+	systemtest.Elasticsearch.ExpectDocs(t, "apm-*-error", nil)
+	expectLogMessage(t, srv, "No Sourcemap available")
+}
+
+func uploadSourcemap(t *testing.T, srv *apmservertest.Server, sourcemapFile, bundleFilepath, serviceName, serviceVersion string) {
+	t.Helper()
+
+	req := newUploadSourcemapRequest(t, srv, sourcemapFile, bundleFilepath, serviceName, serviceVersion)
+	resp, err := http.DefaultClient.Do(req)
+	require.NoError(t, err)
+	defer resp.Body.Close()
+
+	respBody, err := ioutil.ReadAll(resp.Body)
+	require.NoError(t, err)
+	require.Equal(t, http.StatusAccepted, resp.StatusCode, string(respBody))
+}
+
+func newUploadSourcemapRequest(t *testing.T, srv *apmservertest.Server, sourcemapFile, bundleFilepath, serviceName, serviceVersion string) *http.Request {
+	t.Helper()
+
+	var data bytes.Buffer
+	mw := multipart.NewWriter(&data)
+	require.NoError(t, mw.WriteField("service_name", serviceName))
+	require.NoError(t, mw.WriteField("service_version", serviceVersion))
+	require.NoError(t, mw.WriteField("bundle_filepath", bundleFilepath))
+
+	f, err := os.Open(sourcemapFile)
+	require.NoError(t, err)
+	defer f.Close()
+	sourcemapFileWriter, err := mw.CreateFormFile("sourcemap", filepath.Base(sourcemapFile))
+	require.NoError(t, err)
+	_, err = io.Copy(sourcemapFileWriter, f)
+	require.NoError(t, err)
+	require.NoError(t, mw.Close())
+
+	req, _ := http.NewRequest("POST", srv.URL+"/assets/v1/sourcemaps", &data)
+	req.Header.Add("Content-Type", mw.FormDataContentType())
+	return req
+}
+
+func expectLogMessage(t *testing.T, srv *apmservertest.Server, message string) {
+	timeout := time.After(time.Minute)
+	logs := srv.Logs.Iterator()
+	defer logs.Close()
+	for {
+		select {
+		case entry := <-logs.C():
+			if strings.Contains(entry.Message, message) {
+				return
+			}
+		case <-timeout:
+			t.Fatal("timed out waiting for log message")
+		}
+	}
+}
+
+func deleteIndex(t *testing.T, name string) {
+	resp, err := systemtest.Elasticsearch.Indices.Delete([]string{name})
+	require.NoError(t, err)
+	resp.Body.Close()
+	resp, err = systemtest.Elasticsearch.Indices.Flush()
+	require.NoError(t, err)
+	resp.Body.Close()
+}
+
+func expectSourcemapUpdated(t *testing.T, result estest.SearchResult) {
+	for _, hit := range result.Hits.Hits {
+		source := Source{}
```

Comment:
```suggestion
		var source struct {
			Error Error
		}
```

Then we can get rid of the Source type, which has too general a name.